### ISSUE TITLE
Error when exporting products with no condition on completeness

### ISSUE THREAD

**Issue by rainmikko** [akeneo/pim-community-dev#7471]
**Summary**
I found a bug when I tried to export products from Akeneo with no condition on completeness. Bug is reproducable also in [Akeneo demo](http://demo.akeneo.com). I used fresh Akeneo with no modifications and with icecat demo content.

Akeneo version: 2.0.11

**Steps to reproduce**
1. Create export profile with job "Product export in XSLX"
1. On "Content"-tab, change `Completeness` to `No condition on completeness` and Save profile
1. Click "Export now"-button

**Expected result**
Export is finished successfully and exported content is downloadable.

**Actual result**
Export job fails in "Product export"-step after reading product values. Exeption is thrown on backend and command exited with error code "1".
When I runned job at command line in debug mode, it generated following output:
```bash
$ bin/console --env=prod akeneo:batch:job test 21 --env=dev -vvv
15:42:27 DEBUG     [batch] Job execution starting: startTime=, endTime=, updatedTime=, status=2, exitStatus=, exitDescription=[], job=[test]
[]
[]
15:42:27 DEBUG     [batch] Job execution starting: startTime=, endTime=, updatedTime=, status=2, exitStatus=, exitDescription=[], job=[test]
[]
[]
15:42:27 INFO      [batch] Step execution starting: id=0, name=[export], status=[2], exitCode=[EXECUTING], exitDescription=[]
[]
[]
15:42:27 INFO      [batch] Step execution starting: id=0, name=[export], status=[2], exitCode=[EXECUTING], exitDescription=[]
[]
[]
15:42:27 ERROR     [console] Error thrown while running command "--env=prod akeneo:batch:job test 21 --env=dev -vvv". Message: "Return value of Pim\Component\Catalog\Model\ProductModel::getValuesForVariation() must implement interface Pim\Component\Catalog\Model\ValueCollectionInterface, null returned"
[
  "error" => TypeError {
    #message: "Return value of Pim\Component\Catalog\Model\ProductModel::getValuesForVariation() must implement interface Pim\Component\Catalog\Model\ValueCollectionInterface, null returned"
    #code: 0
    #file: "./vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/ProductModel.php"
    #line: 133
    trace: {
      ./vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/ProductModel.php:133 { …}
      ./var/cache/prod/doctrine/orm/Proxies/__CG__PimComponentCatalogModelProductModel.php:254 {
        › 
        ›     return parent::getValuesForVariation();
        › }
      }
      ./vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/VariantProduct.php:129 { …}
      ./vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/VariantProduct.php:80 { …}
      ./vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Converter/MetricConverter.php:45 { …}
      ./vendor/akeneo/pim-community-dev/src/Pim/Component/Connector/Reader/Database/ProductReader.php:98 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Step/ItemStep.php:107 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Step/AbstractStep.php:100 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:263 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:226 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:160 { …}
      ./vendor/akeneo/pim-community-dev/src/Akeneo/Bundle/BatchBundle/Command/BatchCommand.php:155 { …}
      ./vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:252 { …}
      ./vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:954 { …}
      ./vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:83 { …}
      ./vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:240 { …}
      ./vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:71 { …}
      ./vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:148 { …}
      ./bin/console:30 { …}
    }
  },
  "command" => "--env=prod akeneo:batch:job test 21 --env=dev -vvv",
  "message" => "Return value of Pim\Component\Catalog\Model\ProductModel::getValuesForVariation() must implement interface Pim\Component\Catalog\Model\ValueCollectionInterface, null returned"
]
[]
15:42:27 DEBUG     [console] Command "--env=prod akeneo:batch:job test 21 --env=dev -vvv" exited with code "1"
[
  "command" => "--env=prod akeneo:batch:job test 21 --env=dev -vvv",
  "code" => 1
]
[]

In ProductModel.php line 133:
                                                                                                                                                                                              
  [Symfony\Component\Debug\Exception\FatalThrowableError]                                                                                                                                     
  Type error: Return value of Pim\Component\Catalog\Model\ProductModel::getValuesForVariation() must implement interface Pim\Component\Catalog\Model\ValueCollectionInterface, null returned  
                                                                                                                                                                                              

Exception trace:
 Pim\Component\Catalog\Model\ProductModel->getValuesForVariation() at /usr/akeneo/pim-community-standard/var/cache/prod/doctrine/orm/Proxies/__CG__PimComponentCatalogModelProductModel.php:254
 Proxies\__CG__\Pim\Component\Catalog\Model\ProductModel->getValuesForVariation() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/VariantProduct.php:129
 Pim\Component\Catalog\Model\VariantProduct->getAllValues() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Model/VariantProduct.php:80
 Pim\Component\Catalog\Model\VariantProduct->getValues() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Pim/Component/Catalog/Converter/MetricConverter.php:45
 Pim\Component\Catalog\Converter\MetricConverter->convert() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Pim/Component/Connector/Reader/Database/ProductReader.php:98
 Pim\Component\Connector\Reader\Database\ProductReader->read() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Step/ItemStep.php:107
 Akeneo\Component\Batch\Step\ItemStep->doExecute() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Step/AbstractStep.php:100
 Akeneo\Component\Batch\Step\AbstractStep->execute() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:263
 Akeneo\Component\Batch\Job\Job->handleStep() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:226
 Akeneo\Component\Batch\Job\Job->doExecute() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Component/Batch/Job/Job.php:160
 Akeneo\Component\Batch\Job\Job->execute() at /usr/akeneo/pim-community-standard/vendor/akeneo/pim-community-dev/src/Akeneo/Bundle/BatchBundle/Command/BatchCommand.php:155
 Akeneo\Bundle\BatchBundle\Command\BatchCommand->execute() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:252
 Symfony\Component\Console\Command\Command->run() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:954
 Symfony\Component\Console\Application->doRunCommand() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:240
 Symfony\Component\Console\Application->doRun() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:71
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /usr/akeneo/pim-community-standard/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:148
 Symfony\Component\Console\Application->run() at /usr/akeneo/pim-community-standard/bin/console:30

akeneo:batch:job [-c|--config CONFIG] [--username USERNAME] [--email EMAIL] [--no-log] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command> <code> [<execution>]
```

Batch log for runned job looks like following:
```
[2018-01-10 15:42:27] batch.DEBUG: Job execution starting: startTime=, endTime=, updatedTime=, status=2, exitStatus=, exitDescription=[], job=[test] [] []
[2018-01-10 15:42:27] batch.INFO: Step execution starting: id=0, name=[export], status=[2], exitCode=[EXECUTING], exitDescription=[] [] []
```

**Comment by pierallard** [comment]
Hi @rainmikko 

First, thank you for raising this issue!
Second, thank you for beeing extra clear on the description of your issue, it's very pleasant to have message like this !

So I just reproduced the bug on demo.akeneo.com and confirm you it's not the expected behavior. I registered it to our internal bugtracker, and it should be prioritized (without any fix delivery time) under the name `SDS-2520`.

I let the ticket open for let you know when it will be fixed. If you think you can take a look and try to make a PR, we're here to help you doing it, feel free to contribute!

Best regards
Pierre

**Comment by rmatteson** [comment]
Hello! I am getting this issue when exporting products on version 2.1.2, however in my case it occurs regardless of what value is selected under `Completeness`. It does not occur on all product models, if I filter by specific categories certain products fail and certain products succeed. I am still trying to track down why. Any ideas or workarounds?

**Comment by akosglue** [comment]
Hello,

Any updates about this?

**Comment by rmatteson** [comment]
I am willing to help debug this issue, but am pretty unfamiliar with symfony as I am a rails dev. If someone could point me in the right direction to find out what specific product is breaking the export I would be grateful. It's important to my team to get this up and working soon.

**Comment by akosglue** [comment]
This error broke my full export and there was not even a single error message in the process tracker. Only the status that it was failed.
I added a dirty hotfix at the line around 133 in ProductModel.php:
`if($this->code == ''){
$this->values = new ValueCollection();
$this->categories = new ArrayCollection();
$this->products = new ArrayCollection();
$this->productModels = new ArrayCollection();
}
return $this->values;`

After that the export was successful.
I suspect that some products/product models are 'broken', but I haven't found them yet.

**Comment by LaureBro** [comment]
Hi @rainmikko , this issue is fixed on the latest 2.0 (2.0.17), this fix is being reported on higher versions (namely the 2.1). We'll let know you when it's fixed on 2.1 too (cc @rmatteson).
Thanks!

**Comment by kevinrademan** [comment]
I've also reproduced this issue on version 2.2.4. 
@LaureBro 
I found the following PR for 2.0.17 which "seems" to relate to this issue. 
https://github.com/akeneo/pim-community-dev/commit/89e82cab7d64d379118b450976c819e2d58d8ff5

Could you confirm if this wall the full fix, or is this just part of the fix?

**Comment by shugydw** [comment]
Hi @LaureBro , I face this issue with CSV export on 2.1.9. Can you let me know when it would be fixed on 2.1 ?
Thanks!

**Comment by LaureBro** [comment]
Hello!
@shugydw, this issue should have been fixed in 2.1.7. Since the 2.1 is not supported anymore, I'm afraid we won't be able to provide with any further fixes on this version :(
Is it possible for you to migrate on 2.3 version? You should not experience this issue anymore.

Best, 
Laure

**Comment by jmleroux** [comment]
Closed for inactivity
